<launch>
    <!-- nodelet -->
    <arg name="launch_manager" default="true" />
    <arg name="nodelet_manager" default="manager" />
    <arg name="nodelet_manager_namespace" default="dlib_face" />
    <!-- parameters -->
    <arg name="use_detection" default="true" />
    <arg name="queue_size" default="100" />
    <arg name="hz" default="10" />
    <arg name="control_hz" default="false" />
    <arg name="k" default="3" />
    <arg name="threshold" default="0.6" />
    <arg name="model_path" default="$(env HOME)/.ros/dlib/models/mmod_human_face_detector.dat" />
    <arg name="shape_prediction_model_path" default="$(env HOME)/.ros/dlib/models/shape_predictor_5_face_landmarks.dat" />
    <arg name="recognition_model_path" default="$(env HOME)/.ros/dlib/models/dlib_face_recognition_resnet_model_v1.dat" />
    <arg name="face_data_path" default="$(env HOME)/.ros/dlib/face_data" />
    <!-- remaps -->
    <arg name="image" default="image" />
    <arg name="faces" default="face_detection/faces" />
    <arg name="save_face" default="face_recognition/save_face" />

    <group ns="$(arg nodelet_manager_namespace)">

        <node if="$(arg launch_manager)" pkg="nodelet" type="nodelet" name="$(arg nodelet_manager)" args="manager" output="screen" launch-prefix="nice -n 20" />

        <node if="$(arg use_detection)" pkg="nodelet" type="nodelet" name="face_detection" args="load dlib_face/face_detection_nodelet $(arg nodelet_manager)" output="screen" launch-prefix="nice -n 20">
            <param name="queue_size" value="$(arg queue_size)" />
            <param name="hz" value="$(arg hz)" />
            <param name="control_hz" value="$(arg control_hz)" />
            <param name="model_path" value="$(arg model_path)" />
            <remap from="image" to="$(arg image)" />
        </node>

        <node pkg="nodelet" type="nodelet" name="face_recognition" args="load dlib_face/face_recognition_nodelet $(arg nodelet_manager)" output="screen" launch-prefix="nice -n 20">
            <param name="queue_size" value="$(arg queue_size)" />
            <param name="hz" value="$(arg hz)" />
            <param name="control_hz" value="$(arg control_hz)" />
            <param name="k" value="$(arg k)" />
            <param name="threshold" value="$(arg threshold)" />
            <param name="shape_prediction_model_path" value="$(arg shape_prediction_model_path)" />
            <param name="recognition_model_path" value="$(arg recognition_model_path)" />
            <param name="face_data_path" value="$(arg face_data_path)" />
            <remap from="image" to="$(arg image)" />
            <remap from="faces" to="$(arg faces)" />
        </node>

        <node name="save_face_node" pkg="dlib_face" type="save_face.py" output="screen" respawn="true">
            <remap from="save_face" to="$(arg save_face)" />
        </node>
    </group>
</launch>
